contracts/staking/StakingPool.sol

```diff
@@ -0,0 +1,489 @@
+        require(info.totalStakedAmount == 0, "StakePool: totalStakedAmount"); // or should we move this to be a single field?
+
+        if (info.launchPaused) {
+            _pause();
+        }
+
+        _stakingPoolInfo = info;
+    }
+
+    function emergencyWithdraw() external emergencyModeEnabled {
+        _withdrawWithoutRewards();
+    }
+
+    function initializeRewardTokens(
+        address treasury,
+        StakingPoolLib.RewardToken[] calldata rewards
+    ) external atLeastDaoMeepleRole(_stakingPoolInfo.daoId) {
+        _initializeRewardTokens(treasury, rewards);
+    }
+
+    function adminEmergencyRewardSweep()
+        external
+        atLeastDaoAminRole(_stakingPoolInfo.daoId)
+        emergencyModeEnabled
+    {
+        _adminEmergencyRewardSweep();
+    }
+
+    function enableEmergencyMode()
+        external
+        atLeastDaoAminRole(_stakingPoolInfo.daoId)
+    {
+        _stakingPoolInfo.emergencyMode = true;
+    }
+
+    function setRewardsAvailableTimestamp(uint32 timestamp)
+        external
+        atLeastDaoAminRole(_stakingPoolInfo.daoId)
+    {
+        _setRewardsAvailableTimestamp(timestamp);
+    }
+
+    function computeFloatingRewardsPerShare(uint256 rewardTokenIndex)
+        external
+        view
+        returns (uint256)
+    {
+        return _computeFloatingRewardsPerShare(rewardTokenIndex);
+    }
+
+    function currentExpectedReward(address user)
+        external
+        view
+        returns (uint256[] memory)
+    {
+        User memory _user = _users[user];
+        StakingPoolLib.Data memory _info = _stakingPoolInfo;
+
+        uint256[] memory rewards = new uint256[](_info.rewardTokens.length);
+
+        for (uint256 i = 0; i < _info.rewardTokens.length; i++) {
+            if (_info.poolType == StakingPoolLib.StakingPoolType.FLOATING) {
+                rewards[i] = uint256(
+                    (_info.rewardTokens[i].rewardAmountRatio *
+                        _user.depositAmount) / 1 ether
+                );
+            } else {
+                rewards[i] = _computeFixedRewards(_user.depositAmount, i);
+            }
+        }
+        return rewards;
+    }
+
+    function computeFixedRewards(address recipient, uint256 rewardTokenIndex)
+        external
+        view
+        returns (uint128)
+    {
+        return
+            _computeFixedRewards(
+                _users[recipient].depositAmount,
+                rewardTokenIndex
+            );
+    }
+
+    function isRedeemable() external view returns (bool) {
+        //slither-disable-next-line timestamp
+        return _isRewardsAvailable() && _isStakingPeriodComplete();
+    }
+
+    function isRewardsFinalized() external view returns (bool) {
+        return _isRewardsAvailable();
+    }
+
+    function isStakingPeriodComplete() external view returns (bool) {
+        return _isStakingPeriodComplete();
+    }
+
+    function _initializeRewardTokens(
+        address treasury,
```

Comment:
this should probably not be called `treasury`

Reply:
since it might not specifically be a treasury

Reply:
`recipient`?